ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -711,9 +714,51 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 	/* If class/type not defined, match from mechanism */
 	if (get_class(temp) == PKCS11_UNDEFINED_ID &&
```

Comment:
I did quick visual comparison to what we have in our code base for this PR and it seems that this needs some refactoring here about how "guessing" the intent of caller is resolved -- I can try to bake a PR for that.

However I do not see that the change will affect the outcome for this PR but will affect the ones coming later.

Reply:
@vesajaaskelainen , I would agree that guessing part can be cleaned up here. We decided to leave as it is for now and work on it in another PR. If you have the changes available at your end, please feel free to raise a PR for same once this PR is merged.

Reply:
I am OK having that after this PR. We have tested our changes with quite a good set of tools and so far the updated logic seems to be OK. Let's finetune the feature in future PRs while we go forward. In meanwhile I try to rebase all of our changes to track upstream state to make it easier to collaborate.

Reply:
Thanks @vesajaaskelainen . Can you please also help review this and other PR's associated with Key generation in optee_client and optee_test repos.

Reply:
A this stage, these 2 mechanisms mandate the generated object class and type.
So I think here we should first guess the object type (secret key gen. =>CKK_GEN._SECRET, aes key gen.=>CKK_AES, otherwise, CKK_UNDEFINED) and change `sanitize_consistent_class_and_type()` to get 2 new input arguments: class and type, possibly `PKCS11_CKx_UNDEFINED`. The function current use this later value as default class and type when first finding these attributes.

		case PKCS11_CKM_GENERIC_SECRET_KEY_GEN:
		goto out;				class = PKCS11_CKO_SECRET_KEY;
			type = PKCS11_CKK_GENERIC_SECRET;
			break;
		case PKCS11_CKM_AES_KEY_GEN:
			class = PKCS11_CKO_SECRET_KEY;
			type = PKCS11_CKK_AES;

Reply:
I agree with you @etienne-lms . Do you want me to make this change as part of this PR or a separate PR ?

Reply:
i think you can add to this P-R.

Reply:
I have code snippets already for the guessing part so that same semantics work for different use cases too. Do you want that I take this PR and add commit on top of that so you can see how it looks?

Reply:
You can create a P-R (even if it duplicates this one) or simply share a github commit url on your tree, as you like.

Reply:
I will try to port more commits to upstream tree, but here are the changes for this:
https://github.com/vesajaaskelainen/optee_os/commit/91f895166b500c3cdf1b105cc59eeb40eda603e8

I did this on top of @ruchi393's work and whole branch is here:
https://github.com/vesajaaskelainen/optee_os/commits/pkcs11-fix-mecha-attr-contribution

Example snippet for EC and RSA:
```
@@ -692,9 +692,11 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
                        type = PKCS11_CKK_AES;
                        break;
                case PKCS11_CKM_EC_KEY_PAIR_GEN:
+                       class = template_class;
                        type = PKCS11_CKK_DH;
                        break;
                case PKCS11_CKM_RSA_PKCS_KEY_PAIR_GEN:
+                       class = template_class;
                        type = PKCS11_CKK_RSA;
                        break;
                default:
```

```
@@ -421,7 +422,8 @@ enum pkcs11_rc entry_generate_key_pair(struct pkcs11_client *client,
        rc = create_attributes_from_template(&pub_head,
                                             template, template_size, NULL,
                                             PKCS11_FUNCTION_GENERATE_PAIR,
-                                            proc_params->id);
+                                            proc_params->id,
+                                            PKCS11_CKO_PUBLIC_KEY);
        if (rc)
                goto out;
 
@@ -442,7 +444,8 @@ enum pkcs11_rc entry_generate_key_pair(struct pkcs11_client *client,
        rc = create_attributes_from_template(&priv_head,
                                             template, template_size, NULL,
                                             PKCS11_FUNCTION_GENERATE_PAIR,
-                                            proc_params->id);
+                                            proc_params->id,
+                                            PKCS11_CKO_PRIVATE_KEY);
        if (rc)
                goto out;
 
@@ -844,7 +847,8 @@ enum pkcs11_rc entry_derive_key(struct pkcs11_client *client,
        rc = create_attributes_from_template(&head, template, template_size,
                                             parent_obj->attributes,
                                             PKCS11_FUNCTION_DERIVE,
-                                            proc_params->id);
+                                            proc_params->id,
+                                            PKCS11_CKO_UNDEFINED_ID);
        if (rc)
                goto out;
```

Reply:
I was able to make "ugly" version running in "upstream" tree that is now in github:
https://github.com/vesajaaskelainen/optee_os/tree/ugly-tree-starting-point

It is a bit mixed tree and does not follow upstream but is combination of @etienne-lms's tree and then EC improvement support patch by Ricardo Salveti and then lots of fixes while testing with different tools & suites and by reading the specs. There is still work ahead to pass test suites. So far the tools are happy (except the trust policy part of NSS's certutil -- which is rejected)

I am happy to assist in getting things in place but perhaps this PR is not the place. Feel free to propose method.

I also started writing up wiki page how one can test things:
https://github.com/vesajaaskelainen/optee_os/wiki/Testing-thigs

Reply:
@etienne-lms + @ruchi393 : shall we create a issue for integrating pkcs11 support in so that we can discuss there more in details -- eg. what I can take from my tree and so on?

Reply:
Sure, that sounds good. We have our Monthly OP-TEE meeting planned on Monday (Dec 21). If you are planning to join that meeting, we can use that as starting point for our discussion.

At high level, I am currently working to upstream the code in the mainline optee-os by using sks branch from etienne's tree in the order given below :
1. C_GenerateKey (this PR)
2. I have started work on the remaining Object API's - C_FindObjects*, C_GetAttribute, C_CopyObject
3. Next  in plan is C_GenerateKeyPair (RSA)

It would be good to collaborate and work in parallel on this stuff to get things in mainstream sooner.

Reply:
I did not know that such meeting exists. I believe I can join for online meeting on next Monday. Please do send the details to my email.

Reply:
You can find the details on:
https://www.linaro.org/projects/#LOC

Meeting detail is available at 
https://www.trustedfirmware.org/meetings/